File: 2048_8bit/core/src/com/nopalsoft/dosmil/game_objects/BoardPiece.java
package com.nopalsoft.dosmil.game_objects;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.actions.Actions;
import com.nopalsoft.dosmil.Assets;

import java.util.LinkedHashMap;

public class BoardPiece extends Actor {
    public boolean justChanged = false;

    // Positions start counting from left to right from top to bottom.
    final static LinkedHashMap<Integer, Vector2> positionsMap = new LinkedHashMap<>();

    static {
        positionsMap.put(0, new Vector2(20, 350));
        positionsMap.put(1, new Vector2(130, 350));
        positionsMap.put(2, new Vector2(240, 350));
        positionsMap.put(3, new Vector2(350, 350));
        positionsMap.put(4, new Vector2(20, 240));
        positionsMap.put(5, new Vector2(130, 240));
        positionsMap.put(6, new Vector2(240, 240));
        positionsMap.put(7, new Vector2(350, 240));
        positionsMap.put(8, new Vector2(20, 130));
        positionsMap.put(9, new Vector2(130, 130));
        positionsMap.put(10, new Vector2(240, 130));
        positionsMap.put(11, new Vector2(350, 130));
        positionsMap.put(12, new Vector2(20, 20));
        positionsMap.put(13, new Vector2(130, 20));
        positionsMap.put(14, new Vector2(240, 20));
        positionsMap.put(15, new Vector2(350, 20));
    }

    final float SIZE = 110;// Final size of the tab
    public int position;

    // The value worth of this BoardPiece
    private int worth;// I made this piece private because when I change its value I also have to change the image of this piece.
    TextureRegion keyframe;

    public BoardPiece(int position, int worth) {
        this.position = position;
        setWidth(SIZE);
        setHeight(SIZE);
        setOrigin(SIZE / 2f, SIZE / 2f);

        setPosition(positionsMap.get(position).x, positionsMap.get(position).y);
        setWorth(worth);

        if (worth != 0) {// If the piece is worth 0, it is a blue square that has nothing.
            setScale(.8f);
            addAction(Actions.scaleTo(1, 1, .25f));
            Gdx.app.log("Se creo pieza en ", position + "");
        }
    }

    public int getWorth() {
        return worth;
    }

    public void setWorth(int worth) {
        this.worth = worth;
        switch (worth) {
            case 2:
                keyframe = Assets.piece2AtlasRegion;
                break;
            case 4:
                keyframe = Assets.piece4AtlasRegion;
                break;
            case 8:
                keyframe = Assets.piece8AtlasRegion;
                break;
            case 16:
                keyframe = Assets.piece16AtlasRegion;
                break;
            case 32:
                keyframe = Assets.piece32AtlasRegion;
                break;
            case 64:
                keyframe = Assets.piece64AtlasRegion;
                break;
            case 128:
                keyframe = Assets.piece128AtlasRegion;
                break;
            case 256:
                keyframe = Assets.piece256AtlasRegion;
                break;
            case 512:
                keyframe = Assets.piece512AtlasRegion;
                break;
            case 1024:
                keyframe = Assets.piece1024AtlasRegion;
                break;
            case 2048:
                keyframe = Assets.piece2048AtlasRegion;
                break;
            case 0:
            default:
                keyframe = Assets.emptyPieceAtlasRegion;
                break;
        }
    }

    @Override
    public void act(float delta) {
        justChanged = false;
        super.act(delta);
    }

    public void moveToPosition(int pos) {
        this.position = pos;
        Gdx.app.log("Move to ", pos + "");
        addAction(Actions.moveTo(positionsMap.get(position).x, positionsMap.get(position).y, .075f));
    }

    @Override
    public void draw(Batch batch, float parentAlpha) {
        batch.draw(keyframe, getX(), getY(), getOriginX(), getOriginY(), getWidth(), getHeight(), getScaleX(), getScaleY(), getRotation());
    }
}

File: Two_Thousand_Fourty_Eight/core/src/com/nopalsoft/dosmil/game_objects/BoardPiece.kt
package com.nopalsoft.dosmil.game_objects

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.actions.Actions
import com.nopalsoft.dosmil.Assets

class BoardPiece(var currentPosition: Int, worth: Int) : Actor() {
    var justChanged = false

    private val size = 110F // Final size of the tab

    private val positionsMap = LinkedHashMap<Int, Vector2>()

    // The value worth of this BoardPiece
    var worth = 0 // I made this piece private because when I change its value I also have to change the image of this piece.
        set(worth) {
            field = worth
            keyframe = when (worth) {
                2 -> Assets.piece2AtlasRegion
                4 -> Assets.piece4AtlasRegion
                8 -> Assets.piece8AtlasRegion
                16 -> Assets.piece16AtlasRegion
                32 -> Assets.piece32AtlasRegion
                64 -> Assets.piece64AtlasRegion
                128 -> Assets.piece128AtlasRegion
                256 -> Assets.piece256AtlasRegion
                512 -> Assets.piece512AtlasRegion
                1024 -> Assets.piece1024AtlasRegion
                2048 -> Assets.piece2048AtlasRegion
                0 -> Assets.emptyPieceAtlasRegion
                else -> Assets.emptyPieceAtlasRegion
            }
        }
    private var keyframe: TextureRegion? = null

    init {
        width = size
        height = size
        setOrigin(size / 2f, size / 2f)


        this.worth = worth

        if (worth != 0) {
            // If the piece is worth 0, it is a blue square that has nothing.
            setScale(.8f)
            addAction(Actions.scaleTo(1f, 1f, .25f))
            Gdx.app.log("Se creo pieza en ", currentPosition.toString())
        }

        positionsMap[0] = Vector2(20f, 350f)
        positionsMap[1] = Vector2(130f, 350f)
        positionsMap[2] = Vector2(240f, 350f)
        positionsMap[3] = Vector2(350f, 350f)
        positionsMap[4] = Vector2(20f, 240f)
        positionsMap[5] = Vector2(130f, 240f)
        positionsMap[6] = Vector2(240f, 240f)
        positionsMap[7] = Vector2(350f, 240f)
        positionsMap[8] = Vector2(20f, 130f)
        positionsMap[9] = Vector2(130f, 130f)
        positionsMap[10] = Vector2(240f, 130f)
        positionsMap[11] = Vector2(350f, 130f)
        positionsMap[12] = Vector2(20f, 20f)
        positionsMap[13] = Vector2(130f, 20f)
        positionsMap[14] = Vector2(240f, 20f)
        positionsMap[15] = Vector2(350f, 20f)

        setPosition(positionsMap[currentPosition]!!.x, positionsMap[currentPosition]!!.y)
    }

    override fun act(delta: Float) {
        justChanged = false
        super.act(delta)
    }

    fun moveToPosition(destinationPosition: Int) {
        currentPosition = destinationPosition
        Gdx.app.log("Move to ", destinationPosition.toString())
        addAction(Actions.moveTo(positionsMap[currentPosition]!!.x, positionsMap[currentPosition]!!.y, .075f))
    }

    override fun draw(batch: Batch, parentAlpha: Float) {
        batch.draw(keyframe!!, x, y, originX, originY, width, height, scaleX, scaleY, rotation)
    }

}
